src/io/flutter/preview/WidgetEditToolbar.java

```diff
@@ -0,0 +1,396 @@
+import com.jetbrains.lang.dart.assists.AssistUtils;
+import com.jetbrains.lang.dart.assists.DartSourceEditException;
+import icons.FlutterIcons;
+import io.flutter.FlutterInitializer;
+import io.flutter.FlutterMessages;
+import io.flutter.dart.FlutterDartAnalysisServer;
+import io.flutter.inspector.InspectorService;
+import io.flutter.inspector.InspectorGroupManagerService;
+import io.flutter.run.FlutterReloadManager;
+import io.flutter.run.daemon.FlutterApp;
+import io.flutter.utils.EventStream;
+import org.dartlang.analysis.server.protocol.FlutterOutline;
+import org.dartlang.analysis.server.protocol.SourceChange;
+import org.jetbrains.annotations.NotNull;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.*;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Toolbar containing all the widget refactor actions.
+ *
+ * This toolbar is extracted out from the OutlineView so that it can be reused
+ * anywhere we want to expose Flutter widget refactors. For example, as a
+ * toolbar of a property editing popop exposed directly in a code editor.
+ */
+public class WidgetEditToolbar {
+  private class QuickAssistAction extends AnAction {
+    private final String id;
+
+    QuickAssistAction(@NotNull String id, Icon icon, String assistMessage) {
+      super(assistMessage, null, icon);
+      this.id = id;
+      messageToActionMap.put(assistMessage, this);
+    }
+
+    @Override
+    public void actionPerformed(@NotNull AnActionEvent e) {
+      sendAnalyticEvent(id);
+      final SourceChange change;
+      synchronized (actionToChangeMap) {
+        change = actionToChangeMap.get(this);
+        actionToChangeMap.clear();
+      }
+      if (change != null) {
+        applyChangeAndShowException(change);
+      }
+    }
+
+    @Override
+    public void update(AnActionEvent e) {
+      final boolean hasChange = actionToChangeMap.containsKey(this);
+      e.getPresentation().setEnabled(hasChange);
+    }
+
+    boolean isEnabled() {
+      return actionToChangeMap.containsKey(this);
+    }
+  }
+
+  private class ExtractMethodAction extends AnAction {
+    private final String id = "dart.assist.flutter.extractMethod";
+
+    ExtractMethodAction() {
+      super("Extract Method...", null, FlutterIcons.ExtractMethod);
+    }
+
+    @Override
+    public void actionPerformed(@NotNull AnActionEvent e) {
+      final AnAction action = ActionManager.getInstance().getAction("ExtractMethod");
+      if (action != null) {
+        final FlutterOutline outline = getWidgetOutline();
+        if (outline != null) {
+          TransactionGuard.submitTransaction(project, () -> {
+            final Editor editor = getCurrentEditor();
+            if (editor == null) {
+              // It is a race condition if we hit this. Gracefully assume
+              // the action has just been canceled.
+              return;
+            }
+            // TODO(jacobr): how recently was this implemented?
```

Comment:
cc @scheglov

Reply:
DAS was updated in https://dart-review.googlesource.com/c/sdk/+/41520/
Look for the history of `DartServerExtractMethodHandler` in IntelliJ, somehow I cannot do it locally.

Reply:
Looks like that analysis server change landed `Feb 14, 2018`.

Reply:
Great. I've addressed the TODO and verified it still works. Not much changes as IntelliJ just selects the whole block anyway so this is what the user sees.
![image](https://user-images.githubusercontent.com/1226812/70078942-8a7f8480-15b8-11ea-9201-ff8f176eb069.png)